compiler/src/dotty/tools/dotc/transform/init/Semantic.scala

```diff
@@ -616,11 +625,14 @@ object Semantic {
               given Trace = trace1
               val cls = target.owner.enclosingClass.asClass
               val ddef = target.defTree.asInstanceOf[DefDef]
-              val env2 = Env(ddef, args.map(_.value).widenArgs)
               // normal method call
-              withEnv(if isLocal then env else Env.empty) {
-                eval(ddef.rhs, ref, cls, cacheResult = true) ++ checkArgs
-              }
+              if isSyntheticApply(meth) then
```

Comment:
I wondered whether it would make sense to `instantiate` a synthetic apply even if `target.hasSource` is false. But if we don't have source for the synthetic apply, then we probably also don't have source for the class constructor, so `instantiate` wouldn't be useful.

Reply:
> I wondered whether it would make sense to instantiate a synthetic apply even if target.hasSource is false. But if we don't have source for the synthetic apply, then we probably also don't have source for the class constructor, so instantiate wouldn't be useful.

Yes, exactly. 

I found one bug in the code above: the outer for `instantiate` is incorrect. I'll come up with a test to show the bug and fix it in this PR.